drivers/include/periph/hwcrypto.h

```diff
@@ -0,0 +1,394 @@
+ * and on inbetween.
+ *
+ * @param[in] dev           the device to initialize
+ * @param[in] hash          desired hash algorithm
+ *
+ * @return                  0 on success
+ * @return                  -1 if hash is not supported
+ * @return                  -2 if device could not be prepared
+ */
+int hwcrypto_hash_init(hwcrypto_t dev, hwcrypto_hash_t hash);
+
+/**
+ * @brief   Update the digest with a block of data.
+ *
+ * This operation may alter the peripheral's state, therefore it should be used
+ * exclusively. A peripheral may become uninitialized after it is powered off
+ * and on inbetween.
+ *
+ * A hardware crypto device not initialized for hashing may produce unexpected
+ * results.
+ *
+ * @param[in] dev           the device to initialize
+ * @param[in] block         block of data to update digest with
+ * @param[in] block_size    number of bytes in data block
+ *
+ * @return                  number of bytes added to digest
+ * @return                  -1 if hash is not supported
+ * @return                  -2 if block_size is incorrect
+ * @return                  -3 if digest could not be updated
+ */
+int hwcrypto_hash_update(hwcrypto_t dev,
+                         const uint8_t* block,
+                         uint32_t block_size);
+
+/**
+ * @brief   Finalize the hash and copy the result to result buffer.
+ *
+ * This operation may alter the peripheral's state, therefore it should be used
+ * exclusively. A peripheral may become uninitialized after it is powered off
+ * and on inbetween.
+ *
+ * A hardware crypto device not initialized for hashing may produce unexpected
+ * results.
+ *
+ * @param[in] dev           the device to initialize
+ * @param[out] cipher       result output buffer
+ * @param[in] result_size   number of bytes to copy from digest to result
+ *
+ * @return                  number of bytes copied
+ * @return                  -1 if hash is not supported
+ * @returned                -2 if @p result_size exceeds hash size
+ */
+int hwcrypto_hash_final(hwcrypto_t dev,
+                        uint8_t* result,
+                        uint32_t result_size);
+
+/**
+ * @brief   Get mutually exclusive access to the hardware crypto peripheral.
+ *
+ * In case the peripheral is busy, this function will block until the
+ * peripheral is available again.
+ *
+ * @param[in] dev           the device to initialize
+ *
+ * @return                  0 on success
+ * @return                  -1 on error
+ */
+int hwcrypto_acquire(hwcrypto_t dev);
+
+/**
+ * @brief   Release the hardware crypto peripheral to be used by others.
+ *
+ * @param[in] dev           the device to initialize
+ *
+ * @return                  0 on success
+ * @return                  -1 on error
+ */
+int hwcrypto_release(hwcrypto_t dev);
+
+/**
+ * @brief   Power on the hardware crypto peripheral.
+ *
+ * @param[in] dev           the device to initialize
+ */
+void hwcrypto_poweron(hwcrypto_t dev);
```

Comment:
Do you need this function? Does the hwcrypto device ever need to run asynchronously or in the background? Otherwise, I suggest that turning on and off the power should be handled automatically by the other functions

Reply:
Maybe do the power on/off operations in `hwcrypto_acquire`/`hwcrypto_release`?

Reply:
I will now move them to `hwcrypto_acquire`/`hwcrypto_release`.

If we find a use-case than it should be easy to implement them at a later stage.